arangod/RestServer/FlushFeature.cpp

```diff
@@ -20,20 +20,421 @@
 #include "Aql/QueryCache.h"
+#include "Basics/encoding.h"
 #include "Basics/ReadLocker.h"
+#include "Basics/RocksDBUtils.h"
 #include "Basics/WriteLocker.h"
 #include "Cluster/ServerState.h"
 #include "Logger/Logger.h"
+#include "MMFiles/MMFilesDatafile.h"
+#include "MMFiles/MMFilesEngine.h"
+#include "MMFiles/MMFilesLogfileManager.h"
+#include "MMFiles/MMFilesWalMarker.h"
 #include "ProgramOptions/ProgramOptions.h"
 #include "ProgramOptions/Section.h"
 #include "RestServer/DatabaseFeature.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+#include "RocksDBEngine/RocksDBFormat.h"
+#include "RocksDBEngine/RocksDBLogValue.h"
+#include "RocksDBEngine/RocksDBRecoveryHelper.h"
+#include "StorageEngine/EngineSelectorFeature.h"
 #include "Utils/FlushThread.h"
 #include "Utils/FlushTransaction.h"
 
+namespace arangodb {
+
+  /// @brief base class for FlushSubscription implementations
+  class FlushFeature::FlushSubscriptionBase
+      : public FlushFeature::FlushSubscription {
+   public:
+    TRI_voc_tick_t currentTick() const noexcept { return _currentTick; }
+
+   protected:
+    TRI_voc_tick_t _currentTick;
+    TRI_voc_tick_t const _databaseId;
+    arangodb::StorageEngine const& _engine;
+    std::string const _type;
+
+    FlushSubscriptionBase(
+        std::string const& type,
+        TRI_voc_tick_t databaseId,
+        arangodb::StorageEngine const& engine
+    ): _currentTick(engine.currentTick()),
+       _databaseId(databaseId),
+       _engine(engine) {
+    }
+  };
+
+} // arangodb
+
+namespace {
+
+static const std::string DATA_ATTRIBUTE("data"); // attribute inside the flush marker storing custom data body
+static const std::string TYPE_ATTRIBUTE("type"); // attribute inside the flush marker storing custom data type
+
+// wrap vector inside a static function to ensure proper initialization order
+std::unordered_map<std::string, arangodb::FlushFeature::FlushRecoveryCallback>& getFlushRecoveryCallbacks() {
+  static std::unordered_map<std::string, arangodb::FlushFeature::FlushRecoveryCallback> callbacks;
+
+  return callbacks;
+}
+
+arangodb::Result applyRecoveryMarker(
+    TRI_voc_tick_t dbId,
+    arangodb::velocypack::Slice const& slice
+) {
+  if (!slice.isObject()) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      "failed to find a JSON object inside the supplied marker body while applying 'Flush' recovery marker"
+    );
+  }
+
+  if (!slice.hasKey(TYPE_ATTRIBUTE)) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      "failed to find a 'type' attribute inside the supplied marker body while applying 'Flush' recovery marker"
+    );
+  }
+
+  auto typeSlice = slice.get(TYPE_ATTRIBUTE);
+
+  if (!typeSlice.isString()) {
+    return arangodb::Result(
+      TRI_ERROR_INTERNAL,
+      "failed to find a string 'type' attribute inside the supplied marker body while applying 'Flush' recovery marker"
+    );
+  }
+
+  auto type = typeSlice.copyString();
+  auto& callbacks = getFlushRecoveryCallbacks();
```

Comment:
why don't we synchronize access to static variable?
if that's intended, please leave a comment that this particular function is designed to be used in single-threaded environment only

Reply:
please see above